File: Leetcode/min-cost-tickets.py
class Solution:
    
    #Linear Search
    
#     def findNearestGreaterElement(self, days, target):
#         for index, value in enumerate(days):
#             if value > target:
#                 return index
#         #print("Never here!")    
#         return -1    
            
    #BinSearch
    
    def binSearch(self, days, target):
        l = 0
        r = len(days) - 1
        current = math.inf
        currentIndex = -1
        
        while l<=r:
            mid = (l+r) // 2
            
            if days[mid] == target:
                l = mid+1
            elif days[mid] > target:
                r = mid - 1
                if current == math.inf:
                    current = days[mid]
                    currentIndex = mid
                elif days[mid] < current:
                    current = days[mid]
                    currentIndex = mid
            else:
                l = mid+1
            
        return currentIndex   
        
    
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        
        dp = [math.inf for _ in range(0, len(days)+1)]
        
        
        dp[0] = min(costs)
        #print(dp)[1,inf,inf,inf]
        for index, day in enumerate(days):
            if index==0:
                continue
            #print("Index={}".format(index))
            a = costs[0] + dp[index-1]
            #print("a={}".format(a))
            b = None
            if day - 7 <=0:
                #b = costs[1]
                dp[index] = min(costs[1], a)
                #print(dp[index])
            else:
                #_index = self.findNearestGreaterElement(days, day - 7)
                _index = self.binSearch(days, day - 7)
                #print("Index from 7 is:{}".format(_index))
                
                if _index>0:
                    dp[index] = min(dp[_index-1] + costs[1], a)
                else:
                    dp[index] = min(costs[1], a)
            
            if day - 30 < 0:
                dp[index] = min(dp[index], costs[2])
            else:
                #_index = self.findNearestGreaterElement(days, day - 30)
                _index = self.binSearch(days, day - 30)
                #print("Index from 30 is :{}".format(_index))
                
                if _index < 0:
                    dp[index] = min(dp[index], costs[2])
                else:
                    dp[index] = min(dp[_index-1] + costs[2], dp[index])
        #print(dp)        
        return dp[len(days)-1]

File: Leetcode/min-cost-tickets.cpp
class Solution {
public:
    
    // https://leetcode.com/problems/minimum-cost-for-tickets
    
    //Linear Search
//     int search(vector<int> days, int target){
        
//         for(int i=0;i<days.size();i++){            
//             if(days[i]>target) return (i);            
//         }
//         cout<<"Never here"<<endl;
//         return -1;
//     }
    //Binary Search
    int search(vector<int> days, int target){
        
        int l = 0;
        int r = days.size()-1;
        int current = INT_MIN;
        int currentIndex = -1;
        
        while(l<=r){
            
            int mid = (l+r)/2;                        
            
            if(target == days[mid]){
                l = mid+1;
            }
            else if(target > days[mid]){                
                l = mid+1;                
            }
            else{
                if(current==INT_MIN){                    
                    current = days[mid];
                    currentIndex = mid;                    
                }
                else if(days[mid]<current){
                    current = days[mid];
                    currentIndex = mid;
                }
                r = mid-1;
            }
            
        }
        return currentIndex;
    }
    
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        
        vector<int> mem;        
        mem.assign(days.size()+1, INT_MAX-500);
        //mem[0] = costs[0];
        //mem[0] = findMin(costs);        
        
        mem[0] = min(min(costs[0], costs[1]), costs[2]);
        
        
        
        
        for(int i=1;i<days.size();i++){
           
           int a = mem[i-1] + costs[0];
           int b,c;
           if(days[i]-7<=0)               
               mem[i] = min(costs[1], a);
           else{
                                             
               //int ind = search(days, days[i]-7,0, days.size()-1);                          
               int ind = search(days, days[i]-7);                          
               // cout<<"From 7 : index="<<ind<<" days[i] = "<<days[i]<<endl;
               if(ind>0)
                mem[i] = min(mem[ind-1] + costs[1], a);
               else{
                  mem[i]  = min(costs[1], a);
               }
                //mem[i] = min(mem[ind] + costs[1], a);               
           } 
           if(days[i]-30<=0) {
               //mem[i] = min(mem[i-1] + costs[2], mem[i]);
               
               mem[i] = min(costs[2], mem[i]);
           }
           else{                    
               
               //int ind = search(days, days[i]-30, 0, days.size()-1);     
               int ind = search(days, days[i]-30);
               // cout<<"From 30 : index="<<ind<<endl;
               if(ind>0)
                mem[i] = min(mem[ind-1] + costs[2], mem[i]);
               else
                mem[i] = min(mem[ind]+ costs[2], mem[i]);
           }
               
            
        }
        
        // for(int m:mem){
        //     cout<<"mem="<<m<<"\t";
        // }
        // cout<<endl;
        return mem[days.size()-1];
        
    }
};
